### ISSUE TITLE
Adaptive Beta-DPO

### ISSUE THREAD

**Issue by giuseppecurci** [huggingface/trl#5211]
### Feature request

Add **adaptive β strategies** to `DPOTrainer`, starting with **β-DPO (Dynamic β)**.

Paper: https://arxiv.org/pdf/2407.08639 

The proposal is to allow per-batch adaptive scaling of β instead of using a single global value throughout training.

Suggested addition to `DPOConfig`:

```python
adaptive_beta: Optional[str] = None                   # e.g. "beta-dpo"
beta_alpha: Optional[float] = None                     # scaling factor (α)
beta_reference_margin: Optional[float] = None  # optional M0 override
```

When enabled (e.g. `adaptive_beta="beta-dpo"`), β is dynamically adjusted at the batch level:

β_batch = [1 + α (mean(M_batch) − M₀)] β₀

where:
- `β₀` = base DPO β  
- `M_batch` = per-pair margins in the current batch (already computed in the DPO loss)  
- `M₀` = reference margin (running mean or fixed value)

**Why not implement it as a supported `loss_type`?**

As shown in the paper, **β-DPO (Dynamic β)** can be combined with other DPO variants such as SPPO and IPO, since it modifies the scaling parameter β rather than the loss formulation itself.  

It is therefore orthogonal to the specific DPO loss variant and should be implemented as a configurable training strategy (e.g. `adaptive_beta`) instead of a separate `loss_type`. This keeps the design modular and avoids duplicating loss implementations.

### Motivation

In standard DPO, β is fixed across all preference pairs. However, β-DPO shows that optimal β depends on the quality gap between preferred and dispreferred responses:
- Low-gap pairs (ambiguous preferences) benefit from smaller β (stronger updates).
- High-gap pairs (clear preferences) benefit from larger β (more conservative updates).

Real-world datasets contain heterogeneous pair quality and outliers. A single global β can therefore be suboptimal and often requires costly hyperparameter sweeps.

Adaptive β:
- Improves robustness to mixed-quality data
- Reduces sensitivity to β tuning
- Stabilizes training under noisy preference signals

Complexity & Cost
- Compute/Memory overhead: low, compute the average margin and scalar tensor opeartions to compute β_batch 
- No extra models

Fully compatible with current DPO training loop and, as showed in the paper, it can be combined with other losses such as IPO and SPPO. Implementation is localized to the DPO loss and does not affect the rest of the trainer stack.

### Your contribution

I’m willing to open a PR implementing this feature following the contribution guidelines.
The implementation would:
- Add the new config fields
- Modify the DPO loss to support per-batch β
- Include unit tests and documentation updates

Happy to align on design decisions before submitting.

**Comment by giuseppecurci** [comment]
Hi @qgallouedec, I noticed you saw my issue. When you have a moment, could you share your thoughts?

**Comment by qgallouedec** [comment]
Hey @giuseppecurci, sorry for the very slow reply, we're catching up on the backlog.

I read the paper before answering, and Table 1 is what decided it: β-DPO is *two* mechanisms, and the ablation shows the one proposed here (dynamic β alone) is the weaker half (+3.67% / +1.71% / +7.14% on Pythia-410M/1.4B/2.8B, versus +10.84% / +9.84% / +3.71% for β-guided data filtering alone). So `adaptive_beta="beta-dpo"` would ship the smaller share of the paper's gains under the paper's name, and adding the filtering half means per-batch resampling by p(M_i): a data-pipeline change, not a loss-localized one.

It's also less contained than it looks. PR #6123 (thanks @mukund1985) is a fair attempt, and it needed a `use_liger_kernel` incompatibility error because the fused kernel bakes β in https://github.com/huggingface/trl/blob/7073af947351489434f7fd62009bc80212fa1957/trl/trainer/dpo_trainer.py#L860 a new `adaptive_beta_state.json` checkpoint sidecar for the EMA, a DDP gather in the loss hot path, and ~20 substitutions across every loss variant https://github.com/huggingface/trl/blob/7073af947351489434f7fd62009bc80212fa1957/trl/trainer/dpo_trainer.py#L1462-L1586 plus three new public `DPOConfig` fields we'd be supporting more or less forever.

Weighed against one request and no uptake elsewhere, I'd rather not take it on, so I'm closing this and #6123. That said, if you can show the full method (dynamic β *and* β-guided filtering) beating a properly tuned fixed β on a modern 7B+ model, that could change the answer and I'd happily reopen.
